core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
  * @author iyzhang
  */
 public class KernelClient {
+    private static final int MIN_METRIC_POLL_PERIOD_MS = 1000; /* Minimum metrics poll period */
+    private static final int MAX_METRIC_POLL_PERIOD_MS = 512000; /* Maximum metrics poll period */
```

Comment:
I think this is probably too long?  It will take almost 10 minutes to detect a faster or slower link?  I think we should aim for a faster response than that.  Either way, this number needs to be properly considered, and you need to document here why you chose the given number.

Reply:
Since this heartbeat between kernel servers is to just measure latency and transfer rate(i.e., not liveliness), and once we get the stable samples, we can reduce the frequency of this heartbeats gradually.
If the remote kernel server is down due to some reason, anyway, any rpc calls to that particular server fails, then contacts respective DM group policy to get the available server, uses it. And `KernelServerInfo` gets removed from `servers` map after `MAX_FAILED_HEARTBEATS`

Reply:
I think you missed my point.  If a node moves between networks of different speeds (e.g. from wifi to 3G or vice verse) or if it's link simply gets congested, then it will take until the next metric poll period for that change to be detected.  If the metric poll period is 512 seconds, then that's almost 10 minutes before the change gets detected.  That's way too long.

Reply:
Got it. Have reduced it to 128 seconds now.  From the test results, amount of data to send in heartbeats to measure data rate is much negligible when compared to link capacity. So, I believe 128sec(~2min) as upper bound is ok.